common/define-grammar.js

```diff
@@ -261,13 +262,34 @@ module.exports = function defineGrammar(dialect) {
         )
       ),
 
-      function_signature: $ => seq(
+      function_signature: $ => prec.right(seq(
         optional('async'),
         'function',
         field('name', $.identifier),
-        $._call_signature,
-        $._semicolon
-      ),
+        field('type_parameters', optional($.type_parameters)),
+        field('parameters', $.formal_parameters),
+        field('return_type', choice(
+          seq(
+            choice(
+              $.type_annotation,
+              $.asserts,
+              $.type_predicate_annotation
+            ),
+            $._function_signature_semicolon
+          ),
+          seq(
+            $._function_signature_semicolon,
+            optional(
+              choice(
+                $.type_annotation,
+                $.asserts,
+                $.type_predicate_annotation
+              ),
+            ),
+            optional(";")
+          )
+        ))
```

Comment:
I don't understand this change to the grammar. Why does `_function_signature_semicolon` come *before* the type signature?

Reply:
This one was tricky as I had to recreate my train-of-thought from yesterday...

`_function_signature_semicolon` does `lexer->mark_end` at the start and looks far ahead, thus the position in the tree does not matter significantly. Even more so in this case where there are two optional nodes following.

It appears before the type because it can match regardless of the type being there or not. If it were put in front, it would require that a type proceeded it, which is not necessary. Consider:

```
function bar();
```

No type annotation before `;`, but still valid syntax nevertheless.

As to why the other more normal-looking rule is needed despite this, it's to catch the remaining cases where `{` might appear really far forward into the code after a sibling node has already started - which, conversely, means this one must have ended. That sometimes occurs before the chars being detected (`)` and `;`) have appeared (they would have more certainly told that this node has ended). It's not feasible to simply decide between `iswalpha == true` or `{ == false` because it starts matching at the end of `)`, before type annotations *might* appear (types might be matched by `iswalpha` and they can appear on either `function_signature` or `function_declaration`, therefore it's not a feasible heuristic).

Perhaps I can illustrate the situation further by stating that, for detecting that the signature has ended, it's possible to scan for `)`, but not `}`, because the latter might be part of a type annotation (e.g. `function foo(): { foo: bar }`, unlike the former.

I could be missing some more straightforward solution, but this is the one I can reason about right now. I'll take a look later if I can find an example in the wild which parses incorrectly with this approach.